### ISSUE TITLE
Truncated backprop using filtered-transformations

### ISSUE THREAD

**Issue by SNMS95** [patrick-kidger/equinox#954]
Hi everyone,

I am working on doing meta-learning and wanted to implement truncated backprop to estimate the meta-level gradients.
```
import jax.flatten_util
import optax
import equinox as eqx
import equinox.internal as eqxi
import jax
import jax.numpy as jnp

NSTEPS = 500
KIND = 'lax'


def inner_loss(params):
    """Loss function"""
    flat_params = jax.flatten_util.ravel_pytree(params)[0]
    return jnp.sum(flat_params**2)


@eqx.filter_jit
def inner_optimization(params, opt_state, num_steps):

    def inner_opt_step(carry, _):
        """Inner optimization function"""
        params, opt_state = carry
        # Get the gradients
        inner_grads = eqx.filter_grad(inner_loss)(params)
        # Update the parameters
        updates, opt_state = inner_opt.update(inner_grads, opt_state)
        params = eqx.apply_updates(params, updates)
        return (params, opt_state), None

    # make a scan
    init = (params, opt_state)
    (final_params, final_opt_state), _ = eqxi.scan(inner_opt_step, init,
                                                   None, length=num_steps, kind=KIND)
    return final_params, final_opt_state

def truncated_inner_optimization(params, opt_state, num_steps, num_steps_truncated):
    """Truncated inner optimization"""

    @eqx.filter_custom_jvp
    def wrapped_inner_optimization(params, opt_state, num_steps):
        final_params, final_opt_state = inner_optimization(
            params, opt_state, num_steps)
        return final_params, final_opt_state

    @wrapped_inner_optimization.def_jvp
    def wrapped_inner_optimization_jvp(primals, tangents):
        """Truncated inner optimization"""
        primal_out = wrapped_inner_optimization(*primals)
        tangent_out = tangents[:2]
        return primal_out, tangent_out

    # Run the inner optimization with wrapped fn
    final_params, final_opt_state = eqx.filter_jit(
        wrapped_inner_optimization)(params, opt_state, num_steps_truncated)

    # Run the remaining steps
    final_params, final_opt_state = inner_optimization(
        final_params, final_opt_state, num_steps - num_steps_truncated)
    return final_params, final_opt_state

def outer_loss2(params):
    """Outer loss function"""
    # Run optimization
    final_params, _ = truncated_inner_optimization(
        params, inner_opt_state, NSTEPS, NSTEPS - 100)
    # Compute the loss
    flat_final_params = jax.flatten_util.ravel_pytree(final_params)[0]
    return jnp.sum(flat_final_params**3)

# Test
# create a pytree for the parameters 
params = {'w': jax.random.normal(jax.random.PRNGKey(0), (1000, 1000))}
inner_opt = optax.adam(1e-3)
inner_opt_state = inner_opt.init(params)
outer_grads = eqx.filter_grad(outer_loss2)(params)
print(outer_grads)

```

When I checked the memory usage, it seems that this method is working but it feels hacky. 
Is there a better way to do this?

**Comment by patrick-kidger** [comment]
This looks pretty good to me! It's probably worth tweaking things slightly to have a single JIT wrapping the whole thing (including your final `outer_grads = eqx.filter_grad(...`), but otherwise I think this looks about as good as it gets :)

**Comment by SNMS95** [comment]
Hi @patrick-kidger 

This is an even more minimal example. But this errors out. I think it's the `static` argument that's causing  the issues.

The reason I need `static` in the `custom_jvp` is that
in my case, static contains the "meta_parameters". In my original code, I have a single equinox model that contains the meta model and the inner model. I split this model using `eqx.partition` wherever necessary. Thus, if I want only inner parameters, after splitting, I would get params and static. The static contains accumulated gradient information that I need to propagate through.

```
import jax
import jax.numpy as jnp
import equinox as eqx

@eqx.filter_custom_jvp
def f(params, static):
    model = eqx.combine(params, static)
    scale = model(jnp.array([10.0])).sum()
    new_params = jax.tree_map(lambda x: x * scale, params)
    return new_params, static

@f.def_jvp
def f_jvp(primals, tangents):
    params, static = primals
    primal_out = f(params, static)
    return primal_out, tangents

model = eqx.nn.MLP(1, 1, 1, 1, key=jax.random.PRNGKey(0))
params, static = eqx.partition(model, filter_spec=eqx.is_array)

def loss(params, static):
    new_params, _ = f(params, static)
    model = eqx.combine(new_params,
                        static)
    return model(jnp.array([10.0])).sum()

grad = eqx.filter_grad(loss)(params, static)
```

Do you know what should be done to include this?

**Comment by patrick-kidger** [comment]
The output of a `eqx.filter_custom_jvp` must still consist only of JAX types. In this case you're returning the output of `new_inner_model`, which still has its static componets.

(I think)
